src/resolve.ts

```diff
@@ -0,0 +1,47 @@
+import {NonspatialScaleChannel, SCALE_CHANNELS, SPATIAL_SCALE_CHANNELS, SpatialScaleChannel, UNIT_SCALE_CHANNELS} from './channel';
+import * as log from './log';
+import {contains} from './util';
+
+export type ResolveMode = 'independent' | 'shared';
+
+export interface SpatialResolve {
+  scale: ResolveMode;
+  axis?: ResolveMode;
+}
+
+export interface NonspatialResolve {
+  scale: ResolveMode;
+  legend?: ResolveMode;
+}
+
+export type Resolve = SpatialResolve | NonspatialResolve;
+
+export function isSpatialResolve(resolve: Resolve): resolve is SpatialResolve {
+  return 'axis' in resolve;
+}
+
+export function isNonspatialResolve(resolve: Resolve): resolve is NonspatialResolve {
+  return 'legend' in resolve;
+}
+
+export type ResolveMapping = {[P in NonspatialScaleChannel]?: NonspatialResolve} | {[P in SpatialScaleChannel]?: SpatialResolve};
+
+export function initLayerResolve(resolve: ResolveMapping): ResolveMapping {
+  UNIT_SCALE_CHANNELS.forEach(channel => {
```

Comment:
Initializing resolve for all channel no matter they exist or not is weird.  
I think it's better to do this kind of thing in parse when we know if children contains a particular scale or not.  

Also, it seems like this method can cause side effect to the resolve object in the input spec, which is not good.